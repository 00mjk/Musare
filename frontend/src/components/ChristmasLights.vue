<template>
	<div
		:class="{
			'christmas-lights': true,
			loggedIn,
			'christmas-lights-small': small
		}"
	>
		<div class="christmas-wire"></div>
		<template v-for="n in lights" :key="n">
			<span class="christmas-light"></span>
			<div class="christmas-wire"></div>
		</template>
	</div>
</template>

<script>
import { mapState } from "vuex";

export default {
	props: {
		small: { type: Boolean, default: false },
		lights: { type: Number, default: 1 }
	},
	computed: {
		...mapState({
			loggedIn: state => state.user.auth.loggedIn
		})
	},

	async mounted() {
		this.christmas = await lofig.get("siteSettings.christmas");
	}
};
</script>

<style lang="scss" scoped>
.christmas-mode {
	.christmas-lights {
		position: absolute;
		width: 100%;
		height: 50px;
		left: 0;
		top: 64px;
		display: flex;
		justify-content: space-around;
		overflow: hidden;
		pointer-events: none;

		&.christmas-lights-small {
			.christmas-light {
				height: 28px;
				width: 10px;

				&::before {
					width: 10px;
					height: 10px;
				}
			}
		}

		.christmas-light {
			height: 34px;
			width: 12px;
			border-top-left-radius: 50%;
			border-top-right-radius: 50%;
			border-bottom-left-radius: 50%;
			border-bottom-right-radius: 50%;
			z-index: 2;
			box-shadow: 0 2px 3px rgba(10, 10, 10, 0.1),
				0 0 10px rgba(10, 10, 10, 0.1);
			animation: christmas-lights 30s ease infinite;

			&::before {
				content: "";
				display: block;
				width: 12px;
				height: 12px;
				background-color: rgb(6, 49, 19);
				border-top-left-radius: 25%;
				border-top-right-radius: 25%;
				border-bottom-left-radius: 25%;
				border-bottom-right-radius: 25%;
			}

			&:nth-of-type(1) {
				transform: rotate(5deg);
			}

			&:nth-of-type(2) {
				transform: rotate(-7deg);
				animation-delay: -5s;
			}

			&:nth-of-type(3) {
				transform: rotate(3deg);
				animation-delay: -15s;
			}

			&:nth-of-type(4) {
				transform: rotate(10deg);
				animation-delay: -10s;
			}

			&:nth-of-type(5) {
				transform: rotate(-3deg);
				animation-delay: -20s;
			}

			&:nth-of-type(6) {
				transform: rotate(8deg);
				animation-delay: -25s;
			}

			&:nth-of-type(7) {
				transform: rotate(-1deg);
				animation-delay: -30s;
			}

			&:nth-of-type(8) {
				transform: rotate(-4deg);
				animation-delay: -40s;
			}

			&:nth-of-type(9) {
				transform: rotate(3deg);
				animation-delay: -45s;
			}

			&:nth-of-type(10) {
				transform: rotate(-10deg);
				animation-delay: -35s;
			}
		}

		.christmas-wire {
			flex: 1;
			margin-bottom: 15px;
			z-index: 1;

			border-top: 2px solid var(--primary-color);
			border-radius: 50%;
			margin-left: -7px;
			margin-right: -7px;
			transform: scaleY(-1);
			transform-origin: 0% 20%;
		}
	}
}

@keyframes christmas-lights {
	0% {
		background-color: magenta;
	}
	17% {
		background-color: cyan;
	}
	34% {
		background-color: lime;
	}
	51% {
		background-color: yellow;
	}
	68% {
		background-color: orange;
	}
	85% {
		background-color: red;
	}
	100% {
		background-color: magenta;
	}
}
</style>
