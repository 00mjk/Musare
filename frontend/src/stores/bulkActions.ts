import { defineStore } from "pinia";

// TODO fix/decide eslint rule properly
// eslint-disable-next-line
export const useBulkActionsStore = props => {
	const { modalUuid } = props;
	return defineStore(`bulkActions-${modalUuid}`, {
		state: () => ({
			type: null
		}),
		actions: {
			init({ type }) {
				this.type = type;
			}
		}
	})();
};
