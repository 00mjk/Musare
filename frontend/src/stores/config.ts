import { defineStore } from "pinia";

export const useConfigStore = defineStore("config", {
	state: (): {
		config: {
			cookie: string;
			sitename: string;
			recaptcha: {
				enabled: boolean;
				key: string;
			};
			githubAuthentication: boolean;
			messages: Record<string, string>;
			christmas: boolean;
			footerLinks: Record<string, string | boolean>;
			shortcutOverrides: Record<string, any>;
			registrationDisabled: boolean;
			experimental: {
				changable_listen_mode: string[];
				media_session: boolean;
				disable_youtube_search: boolean;
				station_history: boolean;
				soundcloud: boolean;
				spotify: boolean;
			};
		};
	} => ({
		config: {
			cookie: "musareSID",
			sitename: "Musare",
			recaptcha: {
				enabled: false,
				key: ""
			},
			githubAuthentication: false,
			messages: {
				accountRemoval:
					"Your account will be deactivated instantly and your data will shortly be deleted by an admin."
			},
			christmas: false,
			footerLinks: {},
			shortcutOverrides: {},
			registrationDisabled: false,
			experimental: {
				changable_listen_mode: [],
				media_session: false,
				disable_youtube_search: false,
				station_history: false,
				soundcloud: false,
				spotify: false
			}
		}
	}),
	actions: {
		setConfig(config) {
			this.config = config;
		},
		get(query: string) {
			let { config } = this;
			query.split(".").forEach(property => {
				config = config[property];
			});
			return config;
		}
	},
	getters: {
		urls() {
			const { protocol, host } = document.location;
			const secure = protocol !== "http:";
			const client = `${protocol}//${host}`;
			const api = `${client}/backend`;
			const ws = `${secure ? "wss" : "ws"}://${host}/backend/ws`;
			return { client, api, ws, host, secure };
		}
	}
});
