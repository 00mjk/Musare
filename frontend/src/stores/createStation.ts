import { defineStore } from "pinia";

export const useCreateStationStore = ({ modalUuid }: { modalUuid: string }) =>
	defineStore(`createStation-${modalUuid}`, {
		state: () => ({
			official: false
		}),
		actions: {
			init({ official }) {
				this.official = !!official;
			}
		}
	})();
