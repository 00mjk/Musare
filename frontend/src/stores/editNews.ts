import { defineStore } from "pinia";

export const useEditNewsStore = props => {
	const { modalUuid } = props;
	return defineStore(`editNews-${modalUuid}`, {
		state: () => ({
			createNews: false,
			newsId: null,
			sector: "admin"
		}),
		actions: {
			init({ createNews, newsId, sector }) {
				if (createNews) this.createNews = createNews;
				if (newsId) this.newsId = newsId;
				if (sector) this.sector = sector;
			}
		}
	})();
};
