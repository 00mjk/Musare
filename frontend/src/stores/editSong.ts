import { defineStore } from "pinia";
import { ComputedRef, Ref } from "vue";
import { Song } from "@/types/song";
import { Report } from "@/types/report";

export const useEditSongStore = ({ modalUuid }: { modalUuid: string }) =>
	defineStore(`editSong-${modalUuid}`, {
		state: () => ({
			video: {
				player: null,
				paused: true,
				playerReady: false,
				autoPlayed: false,
				currentTime: 0,
				playbackRate: 1
			},
			youtubeId: null,
			song: <Song>{},
			reports: <Report[]>[],
			tab: "discogs",
			newSong: false,
			prefillData: {},
			bulk: false,
			youtubeIds: [],
			songPrefillData: {},
			form: <
				{
					inputs: Ref<{
						[key: string]:
							| {
									value: any;
									originalValue: any;
									validate?: (value: any) => boolean | string;
									errors: string[];
									ref: Ref;
									sourceChanged: boolean;
							  }
							| any;
					}>;
					unsavedChanges: ComputedRef<string[]>;
					save: (saveCb?: () => void) => void;
					setValue: (value: { [key: string]: any }) => void;
					setOriginalValue: (value: { [key: string]: any }) => void;
				}
			>{}
		}),
		actions: {
			init({ song, songs }) {
				if (songs) {
					this.bulk = true;
					this.youtubeIds = songs.map(song => song.youtubeId);
					this.songPrefillData = Object.fromEntries(
						songs.map(song => [
							song.youtubeId,
							song.prefill ? song.prefill : {}
						])
					);
				} else this.editSong(song);
			},
			showTab(tab) {
				this.tab = tab;
			},
			editSong(song) {
				this.newSong = !!song.newSong || !song._id;
				this.youtubeId = song.youtubeId || null;
				this.prefillData = song.prefill ? song.prefill : {};
			},
			setSong(song, reset?: boolean) {
				if (song.discogs === undefined) song.discogs = null;
				this.song = JSON.parse(JSON.stringify(song));
				this.newSong = !song._id;
				this.youtubeId = song.youtubeId;
				const formSong = {
					title: song.title,
					duration: song.duration,
					skipDuration: song.skipDuration,
					thumbnail: song.thumbnail,
					youtubeId: song.youtubeId,
					verified: song.verified,
					artists: song.artists,
					genres: song.genres,
					tags: song.tags
				};
				if (reset && this.form.setValue) this.form.setValue(formSong);
				else if (!reset && this.form.setOriginalValue)
					this.form.setOriginalValue(formSong);
			},
			resetSong(youtubeId) {
				if (this.youtubeId === youtubeId) this.youtubeId = "";
				if (this.song && this.song.youtubeId === youtubeId)
					this.song = {};
			},
			stopVideo() {
				if (this.video.player && this.video.player.pauseVideo) {
					this.video.player.pauseVideo();
					this.video.player.seekTo(0);
				}
			},
			hardStopVideo() {
				if (this.video.player && this.video.player.stopVideo) {
					this.video.player.stopVideo();
				}
			},
			loadVideoById(id, skipDuration) {
				this.song.duration = -1;
				this.video.player.loadVideoById(id, skipDuration);
			},
			pauseVideo(status) {
				if (
					(this.video.player && this.video.player.pauseVideo) ||
					this.video.playVideo
				) {
					if (status) this.video.player.pauseVideo();
					else this.video.player.playVideo();
				}
				this.video.paused = status;
			},
			updateSongField(data) {
				this.song[data.field] = data.value;
			},
			selectDiscogsInfo(discogsInfo) {
				this.song.discogs = discogsInfo;
			},
			updateReports(reports) {
				this.reports = reports;
			},
			resolveReport(reportId) {
				this.reports = this.reports.filter(
					report => report._id !== reportId
				);
			},
			updateYoutubeId(youtubeId) {
				this.song.youtubeId = youtubeId;
				this.loadVideoById(youtubeId, 0);
			},
			updateTitle(title) {
				this.song.title = title;
			},
			updateThumbnail(thumbnail) {
				this.song.thumbnail = thumbnail;
			},
			setPlaybackRate(rate) {
				if (rate) {
					this.video.playbackRate = rate;
					this.video.player.setPlaybackRate(rate);
				} else if (
					this.video.player.getPlaybackRate() !== undefined &&
					this.video.playbackRate !==
						this.video.player.getPlaybackRate()
				) {
					this.video.player.setPlaybackRate(this.video.playbackRate);
					this.video.playbackRate =
						this.video.player.getPlaybackRate();
				}
			}
		}
	})();
