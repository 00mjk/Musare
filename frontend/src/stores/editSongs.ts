import { defineStore } from "pinia";

export const useEditSongsStore = props => {
	const { modalUuid } = props;
	if (!modalUuid || modalUuid === "") return false;
	return defineStore(`editSongs-${modalUuid}`, {
		state: () => ({
			youtubeIds: [],
			songPrefillData: {}
		}),
		actions: {
			init({ songs }) {
				this.youtubeIds = songs.map(song => song.youtubeId);
				this.songPrefillData = Object.fromEntries(
					songs.map(song => [
						song.youtubeId,
						song.prefill ? song.prefill : {}
					])
				);
			}
			// 	resetSongs(state) {
			// 	this.youtubeIds = [];
			// 	this.songPrefillData = {};
			// }
		}
	})();
};
