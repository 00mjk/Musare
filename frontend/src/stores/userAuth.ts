import { defineStore } from "pinia";
import validation from "@/validation";
import ws from "@/ws";
import auth from "@/api/auth";

// TODO fix/decide eslint rule properly
// eslint-disable-next-line
export const useUserAuthStore = defineStore("userAuth", {
	state: () => ({
		userIdMap: {},
		userIdRequested: {},
		pendingUserIdCallbacks: {},
		loggedIn: false,
		role: "",
		username: "",
		email: "",
		userId: "",
		banned: false,
		ban: {},
		gotData: false
	}),
	actions: {
		register(user) {
			return new Promise((resolve, reject) => {
				const { username, email, password } = user;

				if (!email || !username || !password)
					reject(new Error("Please fill in all fields"));
				else if (!validation.isLength(email, 3, 254))
					reject(
						new Error(
							"Email must have between 3 and 254 characters."
						)
					);
				else if (
					email.indexOf("@") !== email.lastIndexOf("@") ||
					!validation.regex.emailSimple.test(email)
				)
					reject(new Error("Invalid email format."));
				else if (!validation.isLength(username, 2, 32))
					reject(
						new Error(
							"Username must have between 2 and 32 characters."
						)
					);
				else if (!validation.regex.azAZ09_.test(username))
					reject(
						new Error(
							"Invalid username format. Allowed characters: a-z, A-Z, 0-9 and _."
						)
					);
				else if (username.replaceAll(/[_]/g, "").length === 0)
					reject(
						new Error(
							"Invalid username format. Allowed characters: a-z, A-Z, 0-9 and _, and there has to be at least one letter or number."
						)
					);
				else if (!validation.isLength(password, 6, 200))
					reject(
						new Error(
							"Password must have between 6 and 200 characters."
						)
					);
				else if (!validation.regex.password.test(password))
					reject(
						new Error(
							"Invalid password format. Must have one lowercase letter, one uppercase letter, one number and one special character."
						)
					);
				else
					auth.register(user)
						.then(res => resolve(res))
						.catch(err => reject(new Error(err.message)));
			});
		},
		login(user) {
			return new Promise((resolve, reject) => {
				auth.login(user)
					.then(() => {
						lofig.get("cookie.SIDname").then(sid => {
							const bc = new BroadcastChannel(
								`${sid}.user_login`
							);
							bc.postMessage(true);
							bc.close();
						});
						resolve({
							status: "success",
							message: "Logged in!"
						});
					})
					.catch(err => reject(new Error(err.message)));
			});
		},
		logout() {
			return new Promise<void>((resolve, reject) => {
				auth.logout()
					.then(() => resolve())
					.catch(() => reject());
			});
		},
		getBasicUser(userId) {
			return new Promise(resolve => {
				if (typeof this.userIdMap[`Z${userId}`] !== "string") {
					if (this.userIdRequested[`Z${userId}`] !== true) {
						this.requestingUserId(userId);
						ws.socket.dispatch(
							"users.getBasicUser",
							userId,
							res => {
								if (res.status === "success") {
									const user = res.data;

									this.mapUserId({
										userId,
										user: {
											name: user.name,
											username: user.username
										}
									});

									this.pendingUserIdCallbacks[
										`Z${userId}`
									].forEach(cb => cb(user));

									this.clearPendingCallbacks(userId);

									return resolve(user);
								}
								return resolve(null);
							}
						);
					} else {
						this.pendingUser({
							userId,
							callback: user => resolve(user)
						});
					}
				} else {
					resolve(this.userIdMap[`Z${userId}`]);
				}
			});
		},
		mapUserId(data) {
			this.userIdMap[`Z${data.userId}`] = data.user;
			this.userIdRequested[`Z${data.userId}`] = false;
		},
		requestingUserId(userId) {
			this.userIdRequested[`Z${userId}`] = true;
			if (!this.pendingUserIdCallbacks[`Z${userId}`])
				this.pendingUserIdCallbacks[`Z${userId}`] = [];
		},
		pendingUser(data) {
			this.pendingUserIdCallbacks[`Z${data.userId}`].push(data.callback);
		},
		clearPendingCallbacks(userId) {
			this.pendingUserIdCallbacks[`Z${userId}`] = [];
		},
		authData(data) {
			this.loggedIn = data.loggedIn;
			this.role = data.role;
			this.username = data.username;
			this.email = data.email;
			this.userId = data.userId;
			this.gotData = true;
		},
		banUser(ban) {
			this.banned = true;
			this.ban = ban;
		},
		updateUsername(username) {
			this.username = username;
		}
	}
});
