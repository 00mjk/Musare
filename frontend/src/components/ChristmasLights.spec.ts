import { mount, flushPromises } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";
import ChristmasLights from "@/components/ChristmasLights.vue";
import { useUserAuthStore } from "@/stores/userAuth";

describe("christmas lights component", () => {
	test("mount", async () => {
		expect(ChristmasLights).toBeTruthy();

		const wrapper = mount(ChristmasLights, {
			global: {
				plugins: [createTestingPinia()]
			}
		});

		expect(wrapper.classes()).toContain("christmas-lights");
		expect(wrapper.html()).toMatchSnapshot();
	});

	test("props", async () => {
		expect(ChristmasLights).toBeTruthy();

		const wrapper = mount(ChristmasLights, {
			global: {
				plugins: [createTestingPinia()]
			},
			props: {
				small: false,
				lights: 1
			}
		});

		expect(wrapper.classes()).not.toContain("christmas-lights-small");
		expect(
			wrapper.findAll(".christmas-lights .christmas-wire").length
		).toBe(1 + 1);
		expect(
			wrapper.findAll(".christmas-lights .christmas-light").length
		).toBe(1);

		await wrapper.setProps({
			small: true,
			lights: 10
		});
		expect(wrapper.classes()).toContain("christmas-lights-small");
		expect(
			wrapper.findAll(".christmas-lights .christmas-wire").length
		).toBe(10 + 1);
		expect(
			wrapper.findAll(".christmas-lights .christmas-light").length
		).toBe(10);

		expect(wrapper.html()).toMatchSnapshot();
	});

	test("loggedIn state", async () => {
		expect(ChristmasLights).toBeTruthy();

		const wrapper = mount(ChristmasLights, {
			global: {
				plugins: [createTestingPinia()]
			}
		});

		const userAuthStore = useUserAuthStore();

		expect(userAuthStore.loggedIn).toEqual(false);
		expect(wrapper.classes()).not.toContain("loggedIn");

		userAuthStore.loggedIn = true;
		await flushPromises();
		expect(userAuthStore.loggedIn).toEqual(true);
		expect(wrapper.classes()).toContain("loggedIn");

		expect(wrapper.html()).toMatchSnapshot();
	});
});
