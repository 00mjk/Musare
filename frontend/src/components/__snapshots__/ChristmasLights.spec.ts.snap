// Vitest Snapshot v1

exports[`christmas lights component > loggedIn state 1`] = `
"<div class=\\"christmas-lights loggedIn\\" data-v-1ff8e05a=\\"\\">
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div>
</div>"
`;

exports[`christmas lights component > mount 1`] = `
"<div class=\\"christmas-lights\\" data-v-1ff8e05a=\\"\\">
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div>
</div>"
`;

exports[`christmas lights component > props 1`] = `
"<div class=\\"christmas-lights christmas-lights-small\\" data-v-1ff8e05a=\\"\\">
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div><span class=\\"christmas-light\\" data-v-1ff8e05a=\\"\\"></span>
  <div class=\\"christmas-wire\\" data-v-1ff8e05a=\\"\\"></div>
</div>"
`;
